dot/core/messages.go

```diff
@@ -32,15 +31,18 @@ func (s *Service) validateTransaction(peerID peer.ID, head *types.Header, rt Run
 	externalExt := types.Extrinsic(append([]byte{byte(types.TxnExternal)}, tx...))
 	validity, err = rt.ValidateTransaction(externalExt)
 	if err != nil {
-		if errors.Is(err, runtime.ErrInvalidTransaction) {
+		logger.Debugf("failed to validate transaction: %s", err)
+		switch err.(type) {
+		case runtime.InvalidTransaction:
 			s.net.ReportPeer(peerset.ReputationChange{
 				Value:  peerset.BadTransactionValue,
 				Reason: peerset.BadTransactionReason,
 			}, peerID)
+			return nil, false, nil
+		case runtime.UnknownTransaction:
+			return nil, false, nil
 		}
```

Comment:
```suggestion
		switch err := err.(type) {
		case runtime.InvalidTransaction:
			s.net.ReportPeer(peerset.ReputationChange{
				Value:  peerset.BadTransactionValue,
				Reason: peerset.BadTransactionReason,
			}, peerID)
			return nil, false, err
		case runtime.UnknownTransaction:
			return nil, false, err
		}
```

Reply:
- Do we want to return the error in the `InvalidTransaction` case? Since it's handled by reporting the peer 🤔 
- Do we want to return the error in the `UnknownTransaction` case? I guess so right?

Reply:
I don't see why not.  The caller should be able to distinguish why the transaction wasn't valid.  I find the middle `bool` parameter a bit unnecessary, since you could probably just do a nil check on the returned `*transaction.Validity`

Reply:
So the handler will stop running if any error is returned. But the code did explicitly keep running if it wasn't a txn validity error, hence why no error was returned and just false. We could return errors and not change code, but then if a transaction is invalid we will stop processing transactions (which as I type it sounds like an attack vector) so don't think we should do that.

Or we can get rid of the bool and do a type check in the caller to determine how to handle. Is this what we want?

Reply:
After discussing in our 1:1. I think removing the bool would make things easier to read and understand.

Reply:
Agreed as well 😉

Reply:
good catch!